builtin/providers/aws/resource_aws_elasticache_replication_group.go

```diff
@@ -0,0 +1,322 @@
+			},
+			"parameter_group_name": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"subnet_group_name": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
+			"security_group_names": &schema.Schema{
+				Type:     schema.TypeSet,
+				Optional: true,
+				Computed: true,
+				Elem:     &schema.Schema{Type: schema.TypeString},
+				Set:      schema.HashString,
+			},
+			"security_group_ids": &schema.Schema{
+				Type:     schema.TypeSet,
+				Optional: true,
+				Computed: true,
+				Elem:     &schema.Schema{Type: schema.TypeString},
+				Set:      schema.HashString,
+			},
+			"engine": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+				Default:  "redis",
+			},
+			"engine_version": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"primary_endpoint": &schema.Schema{
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"preferred_cache_cluster_azs": &schema.Schema{
+				Type:     schema.TypeSet,
+				Optional: true,
+				ForceNew: true,
+				Elem:     &schema.Schema{Type: schema.TypeString},
+				Set:      schema.HashString,
+			},
+		},
+	}
+}
+
+func resourceAwsElasticacheReplicationGroupCreate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).elasticacheconn
+
+	replicationGroupID := d.Get("replication_group_id").(string)
+	description := d.Get("description").(string)
+	cacheNodeType := d.Get("cache_node_type").(string)
+	automaticFailover := d.Get("automatic_failover").(bool)
+	numCacheClusters := d.Get("num_cache_clusters").(int)
+	primaryClusterID := d.Get("primary_cluster_id").(string)
+	engine := d.Get("engine").(string)
+	engineVersion := d.Get("engine_version").(string)
+	securityNameSet := d.Get("security_group_names").(*schema.Set)
+	securityIDSet := d.Get("security_group_ids").(*schema.Set)
+	subnetGroupName := d.Get("subnet_group_name").(string)
+	prefferedCacheClusterAZs := d.Get("preferred_cache_cluster_azs").(*schema.Set)
+
+	securityNames := expandStringList(securityNameSet.List())
+	securityIds := expandStringList(securityIDSet.List())
+	prefferedAZs := expandStringList(prefferedCacheClusterAZs.List())
+
+	req := &elasticache.CreateReplicationGroupInput{
+		ReplicationGroupId:          aws.String(replicationGroupID),
+		ReplicationGroupDescription: aws.String(description),
+		CacheNodeType:               aws.String(cacheNodeType),
+		AutomaticFailoverEnabled:    aws.Bool(automaticFailover),
+		NumCacheClusters:            aws.Int64(int64(numCacheClusters)),
+		PrimaryClusterId:            aws.String(primaryClusterID),
+		Engine:                      aws.String(engine),
+		CacheSubnetGroupName:        aws.String(subnetGroupName),
+		EngineVersion:               aws.String(engineVersion),
+		CacheSecurityGroupNames:     securityNames,
+		SecurityGroupIds:            securityIds,
+		PreferredCacheClusterAZs:    prefferedAZs,
+	}
```

Comment:
if any of these above attributes are omitted because they are optional, this `CreateReplicationGroupInput` will contain empty strings, which is typically not accepted in the API. Does this work when they are omitted?